### ISSUE TITLE
将数据转换为VOC数据集，为什么训练的cls_loss会特别的大

### ISSUE THREAD

**Issue by zhouhing** [zylo117/Yet-Another-EfficientDet-Pytorch#58]
将数据转换为VOC数据集训练，为什么训练的cls_loss会特别的大，voc数据集背景类设为-1，其他类别从0开始
[Info] initializing weights...
Step: 27. Epoch: 0/500. Iteration: 28/2273. Cls loss: 208.66534. Reg loss: 1.27674. Total loss: 209.94209:   1%|          | 28/2273 [00:42<55:17,  1.48s/it]
Process finished with exit code -1

**Comment by zylo117** [comment]
just you wait

**Comment by Cli98** [comment]
> 将数据转换为VOC数据集训练，为什么训练的cls_loss会特别的大，voc数据集背景类设为-1，其他类别从0开始
> [Info] initializing weights...
> Step: 27. Epoch: 0/500. Iteration: 28/2273. Cls loss: 208.66534. Reg loss: 1.27674. Total loss: 209.94209: 1%| | 28/2273 [00:42<55:17, 1.48s/it]
> Process finished with exit code -1

Lol my cls loss start with 37000, after waiting half a day reduced to ~0.65

**Comment by HannH** [comment]
Try to add background class. The loss will decrease normally.
https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/pull/65

**Comment by zylo117** [comment]
> Try to add background class. The loss will decrease normally.
> #65

No, this is wrong. Pretrained efficientdet doesn't have a backgrond class.

Is there a backgrond class in VOC dataset whose index is -1?

Then why don't you set background class index to the end of the class index, like 21?

Then no code needs to be changed.

**Comment by mad-fogs** [comment]
> > 将数据转换为VOC数据集训练，为什么训练的cls_loss会特别的大，voc数据集背景类设为-1，其他类别从0开始
> > [Info] initializing weights...
> > Step: 27. Epoch: 0/500. Iteration: 28/2273. Cls loss: 208.66534. Reg loss: 1.27674. Total loss: 209.94209: 1%| | 28/2273 [00:42<55:17, 1.48s/it]
> > Process finished with exit code -1
> 
> Lol my cls loss start with 37000, after waiting half a day reduced to ~0.65

after half a day, the cls loss comes to be ~0.6 on my dataset.
However, ap50 is 0.02.
what about your datasets?
Well cant even trained a acceptable model by this version.

**Comment by zylo117** [comment]
> > > 将数据转换为VOC数据集训练，为什么训练的cls_loss会特别的大，voc数据集背景类设为-1，其他类别从0开始
> > > [Info] initializing weights...
> > > Step: 27. Epoch: 0/500. Iteration: 28/2273. Cls loss: 208.66534. Reg loss: 1.27674. Total loss: 209.94209: 1%| | 28/2273 [00:42<55:17, 1.48s/it]
> > > Process finished with exit code -1
> > 
> > 
> > Lol my cls loss start with 37000, after waiting half a day reduced to ~0.65
> 
> after half a day, the cls loss comes to be ~0.6 on my dataset.
> However, ap50 is 0.02.
> what about your datasets?
> Well cant even trained a acceptable model by this version.

0.6 is still very high.

for reference, the cls loss of coco with 90 classes is merely 0.18,
but you are using a much easier dataset voc with 20 classes, right? 
so the target clsloss should be much lower

**Comment by mad-fogs** [comment]
> > > > 将数据转换为VOC数据集训练，为什么训练的cls_loss会特别的大，voc数据集背景类设为-1，其他类别从0开始
> > > > [Info] initializing weights...
> > > > Step: 27. Epoch: 0/500. Iteration: 28/2273. Cls loss: 208.66534. Reg loss: 1.27674. Total loss: 209.94209: 1%| | 28/2273 [00:42<55:17, 1.48s/it]
> > > > Process finished with exit code -1
> > > 
> > > 
> > > Lol my cls loss start with 37000, after waiting half a day reduced to ~0.65
> > 
> > 
> > after half a day, the cls loss comes to be ~0.6 on my dataset.
> > However, ap50 is 0.02.
> > what about your datasets?
> > Well cant even trained a acceptable model by this version.
> 
> 0.6 is still very high.
> 
> for reference, the cls loss of coco with 90 classes is merely 0.18,
> but you are using a much easier dataset voc with 20 classes, right?
> so the target clsloss should be much lower

Hi @zylo117 ,
my datasets contains only 4 classes, so the cls loss should be much lower than 0.18 right?
currently it is ~0.14, and the ap50 on val dataset is 0.02...

**Comment by zylo117** [comment]
> > > > > 将数据转换为VOC数据集训练，为什么训练的cls_loss会特别的大，voc数据集背景类设为-1，其他类别从0开始
> > > > > [Info] initializing weights...
> > > > > Step: 27. Epoch: 0/500. Iteration: 28/2273. Cls loss: 208.66534. Reg loss: 1.27674. Total loss: 209.94209: 1%| | 28/2273 [00:42<55:17, 1.48s/it]
> > > > > Process finished with exit code -1
> > > > 
> > > > 
> > > > Lol my cls loss start with 37000, after waiting half a day reduced to ~0.65
> > > 
> > > 
> > > after half a day, the cls loss comes to be ~0.6 on my dataset.
> > > However, ap50 is 0.02.
> > > what about your datasets?
> > > Well cant even trained a acceptable model by this version.
> > 
> > 
> > 0.6 is still very high.
> > for reference, the cls loss of coco with 90 classes is merely 0.18,
> > but you are using a much easier dataset voc with 20 classes, right?
> > so the target clsloss should be much lower
> 
> Hi @zylo117 ,
> my datasets contains only 4 classes, so the cls loss should be much lower than 0.18 right?
> currently it is ~0.14, and the ap50 on val dataset is 0.02...

because the sigmoid function, the slight deviation can lead to big difference.
ie, for D2, coco mAP at cls loss0.18 is 41.5, but at 0.23, mAP is 23.

**Comment by zhouhing** [comment]
我的数据集只有2个类别，昨天训练了一晚上，现在cls_loss一直在1.7浮动，跑了一下评估测试，map只也有0.02+，不知道是不是我voc数据集转换的原因，下一步我打算将数据集转换为coco的格式试一下
![image](https://user-images.githubusercontent.com/51043773/79186688-e999b380-7e4c-11ea-8c57-d933a6b4c7c1.png)

**Comment by zylo117** [comment]
it's a bug, so sorry for the trouble. please pull the latest code and try again

**Comment by zhouhing** [comment]
> 这是一个错误，非常抱歉。请提取最新代码，然后重试

Thank you. I'll try again

**Comment by angryhen** [comment]
> > 这是一个错误，非常抱歉。请提取最新代码，然后重试
> 
> 谢谢。我会再尝试

Have you solved this problem？
The result I got is similar to yours, and it is also from voc to coco. I once suspected that there is a problem with the data conversion. But I cant solve it
